This patch provides GlusterFS CSI driver which is compatible with GD2 (https://github.com/gluster/glusterd2).

humblec opened the pull request:
This patch provides GlusterFS CSI driver which is compatible
    with GD2 (https://github.com/gluster/glusterd2).
    
    This code is capable of volume provision, deletion, mount and unmount against
    CSI spec v0.3.
    The deployment artifacts are delivered by the docker container images.
    The dir structure of code is carved out based on the discussions on
    https://github.com/gluster/gluster-csi-driver/pull/1#issuecomment-370506170
    
    Demo url: https://asciinema.org/a/195024
    
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
@amarts PTAL .

humblec commented:
@madhu-1 PTAL.

Madhu-1 commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,375 @@
+	}
+
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "ValidateVolumeCapabilities() - Volume ID is nil")
```
trivial: update error message to ```volume ID is empty```.

humblec replied:
nil looks better Madhu.

Madhu-1 commented on cmd/glusterfs/main.go:
```diff
@@ -0,0 +1,58 @@
+	if config.Endpoint == "" {
+		config.Endpoint = ep
+	}
+	d := gfd.New(config)
```
we are returning nil from ```gfd.New(config)``` in case of config is ```nil``` do need to handle ```d==nil```?

Madhu-1 commented on README.md:
```diff
@@ -1,3 +1,169 @@
+#kubectl create  -f csi-attacher-glusterfsplugin.yaml
+#kubectl create -f csi-nodeplugin-glusterfsplugin.yaml
+#kubectl create -f csi-provisioner-glusterfsplugin.yaml
+
```
update this to create ```RBAC``` for csi-driver from yaml file

Madhu-1 commented on pkg/glusterfs/driver.go:
```diff
@@ -0,0 +1,62 @@
+	}
+}
+
+//TODO:
```
trivial: remove this TODO

Madhu-1 commented on README.md:
```diff
@@ -1,3 +1,169 @@
+
+### Create a pod with this claim
+~~~
+[root@localhost cluster]# cat ../demo/fedora-pod.json
```
```fedora-pod.json``` file is missing in deploy folder

humblec commented on README.md:
```diff
@@ -1,3 +1,169 @@
+###Deploy a GD2 gluster cluster
+###Deploy CSI driver
+~~~
+#kubectl create  -f csi-attacher-glusterfsplugin.yaml
```
I will open an issue for this and address later.

humblec commented on README.md:
```diff
@@ -1,3 +1,169 @@
+  annotations:
+    storageclass.beta.kubernetes.io/is-default-class: "true"
+provisioner: org.gluster.glusterfs
+[root@localhost cluster]# 
```
removed.

humblec commented:
@Madhu-1 addressed the comments. The rest of the improvements in README I will pull in another PR.

humblec commented:
Thanks @Madhu-1 for your review and also for helping out on verification!!

prashanthpai commented on Gopkg.toml:
```diff
@@ -0,0 +1,50 @@
+
+[[constraint]]
+  name = "github.com/gluster/glusterd2"
+  version = "4.1.0"
```
Don't pin it to that version. Use latest master

humblec replied:
done.

kshlm commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+		} else {
+			resp.Supported = false
+		}
+	}
```
This is buggy. We would be returning `Supported = true` when the last requested capability is supported even if earlier capabilities are not. You should be breaking out when any requested capability is unsupported.

kshlm commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+	}
+	var entries []*csi.ListVolumesResponse_Entry
+	for _, vol := range volumes {
+		v, e := cs.client.VolumeStatus(vol.Name)
```
We should implement a `VolumeStatusAll` API in this. It would simplify this so much and make it perform better at the scale of volumes GCS expects. Filed issue gluster/glusterd2#1136 to track this.

Madhu-1 replied:
if we store volume size during the volume creation, we don't need to implement new API in gd2.
I have created an issue to add volume size filed in volume info.  [issue](https://github.com/gluster/glusterd2/issues/1081)

prashanthpai replied:
@kshlm The `csi.ListVolumesResponse` only needs volume ID and Capacity. All we have to do is add capacity to volume info. We already support listing all volume infos in a single API call

kshlm commented on CHANGELOG.md:
```diff
@@ -0,0 +1,8 @@
+*  Update deployment artifacts.
+   Humble Chirammal hchiramm@redhat.com
+
+# v1.0.0
```
This is just my opinion, but let's not jump to v1 right away. Let's spend some time at v0.y.z before ending up at v1.

kshlm commented on pkg/glusterfs/deploy/kubernetes/csi-attacher-glusterfsplugin.yaml:
```diff
@@ -0,0 +1,81 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
I don't understand how the project is being licensed. We have LGPLv3 and GPLv2 COPYING files added, but these K8S definition files are licensed as Apache2!? I have 2 questions.

1. Why?
2. How is it supposed to work?

humblec replied:
@kshlm as mentioned in my last comments , license part is unclear and we need to reach to conclusion. some discussions were on https://github.com/gluster/gluster-csi-driver/pull/1

kshlm commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+	KB = 1 << (10 * iota)
+	MB
+	GB
+	TB
```
If we have to support human friendly values, I'd prefer we use a well used package instead of rolling our own. The go-humanize package is used widely, and I believe K8S uses it as well to support human friendly values in the definition files.

Madhu-1 commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+	glog.V(2).Infof("Volume list response: %+v", volumeListResp)
+
+	/* TODO: Check the requested volume exist in the cluster
+	glusterServer, bkpServers, err = cs.checkExistingVolumes(&volumeListResp, volumeName, volSizeBytes)
```
in gd2, all peer related api's works with peerID but volume related api's works with the volume name.
we should follow the same mechanism irrespective of API's.
I have created an issue to do all volume related operation based on the volume ID (which is assumed to be unique. [issue](https://github.com/gluster/glusterd2/issues/1107)

once its implemented, CSI driver should list volumes and then do the comparison.

amarts commented:
@humblec thanks for the PR.. Looks good in general, and thanks for the [asciinema ](https://asciinema.org/a/195024), proves it works.

Checked the comments, as they look valid, can you please refresh the patches addressing the few in the code? I am fine with another PR for README.

prashanthpai commented on pkg/glusterfs/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+# Copyright 2018 The Gluster CSI Authors.
```
This has the Apache license where as the project has (L)GPL. Is that intended ?

humblec replied:
same as mentioned in the last comment, license is yet to be sorted out.

prashanthpai replied:
In that case, it's better to stick to one (GPL for now), than to introduce a mix of both. If need be, this can be changed to BSD license elsewhere.

prashanthpai commented on pkg/glusterfs/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM centos:7.4.1708
```
Any specific reason not to use the latest ?

humblec replied:
Added to latest tag

prashanthpai commented on pkg/glusterfs/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+# Copy glusterfs-csi-driver from build directory
+COPY ./build/glusterfs-csi-driver /glusterfs-csi-driver
+
+RUN yum -y install glusterfs-fuse && yum -y install epel-release && yum -y install jq && yum clean all
```
All yum things can be one command to ensure fewer fs layers get built.

The `yum clean all` needs a `-y` flag.

prashanthpai commented on pkg/glusterfs/deploy/kubernetes/csi-provisioner-glusterfsplugin.yaml:
```diff
@@ -0,0 +1,79 @@
+# Copyright 2018 The Gluster CSI Authors.
```
Same comment about the license here

humblec replied:
same as above.

aravindavk commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+const (
+	glusterDescAnn        = "GlusterFS-CSI"
+	glusterDescAnnValue   = "gluster.org/glusterfs-csi"
+	defaultVolumeSizeInMB = 20 * MB // minimum volume size
```
Change the default volume size to 1GB(or meaningful default size). 20MB is minimum size validated by Glusterd2, but here this default size will be used when user don't specify the size.

Madhu-1 replied:
@aravindavk  we need to finalize what should be the default size we should provide.
currently in heketi its `1GB` because minimum required size to create volume is `1GB`

humblec replied:
The constant is renamed and I also use a seperate import to calculate this.

prashanthpai commented on cmd/glusterfs/main.go:
```diff
@@ -0,0 +1,58 @@
+
+	cmd.PersistentFlags().StringVar(&config.RestURL, "resturl", "", "glusterd2 rest endpoint")
+
+	cmd.PersistentFlags().StringVar(&config.RestUser, "username", "glustercli", "glusterd2 user name")
```
We don't have full fledged user management yet. I'd focus on having TLS support as the pods may not run on the same host as where `glusterd2` is running

prashanthpai commented on cmd/glusterfs/main.go:
```diff
@@ -0,0 +1,58 @@
+		os.Exit(1)
+	}
+
+	os.Exit(0)
```
This isn't required. It's the default

prashanthpai commented on cmd/glusterfs/main.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+func handle(config *utils.Config) {
+	ep := os.Getenv("CSI_ENDPOINT")
```
You can use `cobra` library's features to automatically consume and use environment variables.

See https://github.com/gluster/glusterd2/pull/1114 for example usage

prashanthpai commented on pkg/glusterfs/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+
+RUN yum -y install glusterfs-fuse && yum -y install epel-release && yum -y install jq && yum clean all
+
+RUN rm -rf /var/cache/yum
```
This isn't needed, adds an additional layer to the image

aravindavk commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+	glog.V(3).Infof("Request fields:[ %v %v %v %v %v %v]", glusterVol, glusterServer, glusterURL, glusterURLPort, glusterUser, glusterUserSecret)
+
+	// Get list of volumes in the TSP
+	volumeListResp, err := cs.client.Volumes("")
```
Volume list can be skipped here. Try creating Volume, if fails with Exists error then get the volinfo and respond.

(Not for current PR, can be addressed later)

humblec replied:
sure, will get it in next pr.

prashanthpai commented on pkg/glusterfs/driver.go:
```diff
@@ -0,0 +1,62 @@
+import (
+	"github.com/gluster/gluster-csi-drivers/pkg/glusterfs/utils"
+
+	"github.com/gluster/glusterd2/pkg/restclient"
```
Needs a line to separate glusterfs imports from third-party ones

prashanthpai commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+
+	glog.V(2).Infof("Volume list response: %+v", volumeListResp)
+
+	/* TODO: Check the requested volume exist in the cluster
```
What's the expected action if the volume exists ?

Madhu-1 replied:
some more validation as per CSI specification, if all checks are passed we need to send back the volume info

prashanthpai commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+	}
+	*/
+
+	// If volume does not exist, provision volume
```
You can offload checking if volume exists to glusterd2 which will do this anyway. The CSI driver shouldn't be doing this

Madhu-1 replied:
as per CSI spec if the volume is already present we need to return back success

prashanthpai commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+	glog.V(3).Infof("Request fields:[ %v %v %v %v %v %v]", glusterVol, glusterServer, glusterURL, glusterURLPort, glusterUser, glusterUserSecret)
+
+	// Get list of volumes in the TSP
+	volumeListResp, err := cs.client.Volumes("")
```
This is unnecessary if the purpose is only to check if a volume exists

Madhu-1 replied:
if the volume is present when need to send back the success status based on some checks on volume

prashanthpai commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to create volume: %s", err.Error()))
+	}
+
+	glog.V(3).Infof("volume create response : %+v", volumeCreateResp)
```
There should be distinction between info and debug levels. This entry dumps the whole response

prashanthpai commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+
+	for i, p := range peers {
+		if i == 0 {
+			for _, a := range p.PeerAddresses {
```
This should be `ClientAddresses` (same as endpoint)`

Madhu-1 replied:
gd2 runs as a privileged container. but as per default glusterd2 configuration, glusterd2 will listen on all IP addresses, which will have `localhost`,`docker ip ` and other `host ips`  are attached to the container.

we need IP address to mount a volume to app pods.
`PeerAddresses` is having host IP address, which can be used for a mount point.

do we have any use case in running gd2 container on multiple IP?

humblec replied:
Ideally GD2 should have an API which return the brick servers or backup servers or gluster peer addresses. It will make life a lot easier for integrations. @Madhu-1 @prashanthpai can any one open an issue in GD2 for the same ?

prashanthpai replied:
I'll open an issue for the same.

The backup servers are slightly less relevant as the glusterfs client (fuse,libgfapi) will send a get this list in the response to `GETSPEC` that happens during mount.

Refer: https://github.com/gluster/glusterd2/issues/382

prashanthpai replied:
> default glusterd2 configuration, glusterd2 will listen on all IP addresses

The default behaviour will stay. It's configurable though, via the config file or a command line argument and via ENV variables that can be used in containers.

`PeerAddress` should only be used for peer operations - addand remove peer. PeerAddress may go away in future.

Madhu-1 commented on cmd/glusterfs/main.go:
```diff
@@ -0,0 +1,58 @@
+	"fmt"
+	"os"
+
+	gfd "github.com/gluster/gluster-csi-drivers/pkg/glusterfs"
```
need to change this import path `gluster-csi-drivers` to `gluster-csi-driver`

Madhu-1 replied:
currently building is failing for me because of this

humblec replied:
its based on the gopath, any way I renamed to gluster-csi-driver for parity with this repo.

Madhu-1 commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,377 @@
+	}
+
+	glog.V(3).Infof("volume create response : %+v", volumeCreateResp)
+	err = cs.client.VolumeStart(volumeName, true)
```
I have some though on this
I agree with @kshlm, in perspective of resource utilization.

the problem here is during volume mount, what if volume start fails? kubernetes will never try to delete and recreate the volume. it will always try to mount the volume

if we start volume during the volume creation time, if volume start fails we can delete the created volume and return back an error to the kubernets, 
kubernetes will retry to create the PVC because of failure

humblec replied:
@kshlm @Madhu-1 By volume provisioned success, in kube it means its ready to be served. In that sense keeping it activated is the expectation. May be there can be other queries on volume independent of mount call.

humblec commented:
@Madhu-1 @kshlm @aravindavk @prashanthpai  @amarts Thanks for the comments. I will address most in next patch set. One thing which I dont have clue here is the licence part, I have copied glusterfs licence , but I also need to accommodate   apache licence of kube and kube has or need a contributor agreement. so license is vague to me and I am planning to discuss with @JohnStrunk and others on this.

prashanthpai commented:
> I will address most in next patch set. 

Do you mean in a separate PR or do you plan to update this PR ?

humblec commented:
@prashanthpai most of it addressed in the latest patch set. The license part need to be sorted out. The license part discussion we kickstarted here https://github.com/gluster/gluster-csi-driver/pull/1, but I am still vague on it.

humblec commented on Gopkg.toml:
```diff
@@ -0,0 +1,50 @@
+
+[[constraint]]
+  name = "github.com/gluster/glusterd2"
+  version = "4.1.0"
```
Done

humblec commented on README.md:
```diff
@@ -1,3 +1,169 @@
 # gluster-csi-driver
 
 This repo contains CSI driver for Gluster. The Container Storage Interface (CSI) is a proposed new industry standard for cluster-wide volume plugins.  “Container Storage Interface” (CSI)  enables storage vendors (SP) to develop a plugin once and have it work across a number of container orchestration (CO) systems. 
+
```
Glusterd2 has been replaced as GD2. This repo readme is in parent dir so gluster mention is fine as other drivers can also get space in this README.

humblec commented:
@amarts @aravindavk @kshlm @prashanthpai @Madhu-1  I have addressed 95% of the comments (quickly:)) and mainly pending ones are on LICENSE part. PTAL. Thanks for review!

prashanthpai commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,370 @@
+
+	/* TODO: Check the requested volume exist in the cluster
+	glusterServer, bkpServers, err = cs.checkExistingVolumes(&volumeListResp, volumeName, volSizeBytes)
+	if err == nil {
```
Remove commented out code.

prashanthpai commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,370 @@
+	glog.V(3).Infof("volume create response : %+v", volumeCreateResp)
+	err = cs.client.VolumeStart(volumeName, true)
+	if err != nil {
+		//TODO : delete the volume
```
This todo can be handled. It's just a few lines of code and one call to delete volume.

humblec replied:
Would like to cover the negatives in a single commit which is tracked by https://github.com/gluster/gluster-csi-driver/issues/12

Madhu-1 replied:
I think we don't need to delete the volume if volume start fails
kubernetes will retry if volume creation fails as we are listing the volumes and starting it again before sending back the response

humblec commented on CHANGELOG.md:
```diff
@@ -0,0 +1,12 @@
+# v0.0.7
+* Address review comments.
+  Humble Chirammal <hchiramm@redhat.com>
+
```
The changelog version and entries are matching with the docker image tags we use for subject containers, so it make sense.  One additional thing I can do here is, update the version string in the code as soon as there is a changelog entry.

JohnStrunk replied:
I agree that since we don't have any releases, we shouldn't have them here in the changelog w/ versions. There definitely shouldn't be tagged image versions w/o a github release.

I think this file can be collapsed to:
```
# master
- Intital version of CSI driver
```

When we cut a release, we can change `master` to the releaase tag

humblec commented on CHANGELOG.md:
```diff
@@ -0,0 +1,12 @@
+# v0.0.7
```
Updated commit message and PR title.

humblec commented on Gopkg.toml:
```diff
@@ -0,0 +1,50 @@
+
```
Done.

humblec commented on cmd/glusterfs/main.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+func handle(config *utils.Config) {
+	ep := os.Getenv("CSI_ENDPOINT")
```
I think its fine, the var may be useful later.

humblec commented:
@prashanthpai PTAL.

nixpanic commented:
![](https://github.trello.services/images/mini-trello-icon.png) [csi](https://trello.com/c/TVrHD6Ha/10-csi)

kshlm commented:
There is still one obvious [bug](https://github.com/gluster/gluster-csi-driver/pull/11#discussion_r207901049) that is unaddressed. Even though I don't think it would cause much problems in the work of the CSI driver right now, I'd like it to be addressed.

In addition, I'd like the licensing stuff to be clarified as well.

Apart from these, I don't have any other blockers for getting this in.

JohnStrunk commented:
@humblec I'm unable to build this PR (and there are no build instructions that I can find)

My results:
```
$ pwd
/home/jstrunk/go/src/github.com/gluster/gluster-csi-driver
$ git checkout humblec/glusterfs-csi-gd2
HEAD is now at 68f19a35 This patch provides GlusterFS CSI driver which is compatible with GD2 (https://github.com/gluster/glusterd2).
$ make clean
go clean -r -x
cd /home/jstrunk/go/src/github.com/gluster/gluster-csi-driver
rm -f gluster-csi-driver.test gluster-csi-driver.test.exe
rm -rf build
$ rm -rf vendor
$ make
if [ ! -d ./vendor ]; then dep ensure -v; fi
Root project is "github.com/gluster/gluster-csi-driver"
 3 transitively valid internal packages
 10 external packages imported from 7 projects
(0)   ✓ select (root)
(1)	? attempt github.com/container-storage-interface/spec with 1 pkgs; at least 1 versions to try
(1)	    try github.com/container-storage-interface/spec@v0.3.0
(1)	✓ select github.com/container-storage-interface/spec@v0.3.0 w/1 pkgs
(2)	? attempt github.com/golang/protobuf with 2 pkgs; at least 1 versions to try
(2)	    try github.com/golang/protobuf@v1.1.0
(2)	✓ select github.com/golang/protobuf@v1.1.0 w/2 pkgs
(3)	? attempt github.com/gluster/glusterd2 with 2 pkgs; at least 1 versions to try
(3)	    try github.com/gluster/glusterd2@master
(4)	✗   github.com/gluster/glusterd2@master not allowed by constraint master:
(4)	    master from (root)
(3)	    try github.com/gluster/glusterd2@v4.1.0
(4)	✗   github.com/gluster/glusterd2@v4.1.0 not allowed by constraint master:
(4)	    master from (root)
.....
```
Looks like there's something going on w/ the glusterd2 dependency.

amarts commented:
> Also. how does this relate (if at all) to @Madhu-1's work on a GD2 based CSI driver here:
> 
> https://github.com/Madhu-1/glusterd2-csi-driver

This happened because while discussing about GCS work, while both @obnoxxx and @humblec were away, I was discussing with @Madhu-1 about sample GD2 based CSI. Meantime, @humblec had also tried and had some code. I guess this PR is output of both of them agreeing to work on it together.

> How does this PR related to PR #1 which adds a CSI driver based on the current world (heketi + gd1)? If this supersedes the other one, we should probably close #1. But I think we should still decide whether we want to provide a CSI driver for the heketi world.

If I have to vote, I would vote for getting #1 to completion, if it can be done soon (ie, within another 2 weeks). It makes sense as it provides the users with an CSI driver for existing products.

> locked down the IVP api in glusterd2, since that got merged into master / gluster 4.1 with the statement that it is not in its final form and subject to massive change.
> 
> merged the robustness fixes for the current transaction framework of GD2 (that are currently under discussion).

I would say that we should surely 'merge' it, as it gives a possibility for us to have the pieces end to end, and continue the effort to get the testing going.

I am with you in saying, we shouldn't be making a release of it, knowing there are known stability issues. That would be bad.

prashanthpai commented:
> I personally think (as I said many times) that it would be beneficial to have a configurable driver that we can point at heketi or gd2. That would help testing etc. Overhead and duplication should be rather minimal.

This should be easily doable within a reasonable amount of time. I'm not against it.

> Furthermore, I would request this GD2 part not to be merged until we have

> locked down the IVP api in glusterd2, since that got merged into master / gluster 4.1 with the statement that it is not in its final form and subject to massive change.

The API is "open" to changes that improve it. Suggestions and inputs are welcome. Any changes to implementation detail shouldn't matter to CSI driver.

> merged the robustness fixes for the current transaction framework of GD2 (that are currently under discussion).

I believe that work should not block this CSI driver effort because all near-term changes to internal implementation details of the glusterd2 transaction framework won't have any ramifications on glusterd2 APIs being consumed by CSI driver.

In general, CSI driver will greatly help glusterd2 stabilise management operations used by k8s customers as the CSI driver can be used as test driver to extensively exercise a focused subset of glusterd2 APIs and operations.

IMO, considering that the CSI driver is itself a very small shim layer:
* Any pending work in glusterd2 shouldn't block merging this PR.
* We must not defer fixing/addressing comments raised now about contents of this PR.

obnoxxx commented:
> >  Also. how does this relate (if at all) to @Madhu-1's work on a GD2 based CSI driver here:
> >
> >  https://github.com/Madhu-1/glusterd2-csi-driver
> 
> This happened because while discussing about GCS work, while both @obnoxxx and @humblec were away, I was discussing with @Madhu-1 about sample GD2 based CSI. Meantime, @humblec had also tried and had some code. I guess this PR is output of both of them agreeing to work on it together.

That sounds great!

> >  How does this PR related to PR #1 which adds a CSI driver based on the current world (heketi + gd1)? If this supersedes the other one, we should probably close #1. But I think we should still decide whether we want to provide a CSI driver for the heketi world.
> 
> If I have to vote, I would vote for getting #1 to completion, if it can be done soon (ie, within another 2 weeks). It makes sense as it provides the users with an CSI driver for existing products.

Thanks, I'd vote for the same. I don't necessarily have the 2 weeks thing. Apparently, the code merge was to a good extent so far blocked by not having a proper CI infrastructure etc.

Of course, if we agree to have the GD2 flavour as a (configurable) mode of our one CSI driver (which would be my preference), then this PR #11 would need to be reworked on top of #1. But well, that's just life. :-)

> > Furthermore, I would request this GD2 part **not to be merged** until we have
> > * locked down the IVP api in glusterd2, since that got merged into master / gluster 4.1 with the statement that it is not in its final form and subject to massive change.
> > *  merged the robustness fixes for the current transaction framework of GD2 (that are currently under discussion).
> 
> I would say that we should surely 'merge' it, as it gives a possibility for us to have the pieces end to end, and continue the effort to get the testing going.
> 
> I am with you in saying, we shouldn't be making a release of it, knowing there are known stability issues. That would be bad.

Ok, I agree with giving **us** (as in the devs) a way to test is good. My main concern is: If other people start using it, we may have more difficulties changing things later on. :-)

amarts commented:
Posted #15 for license update for the whole repo. Let this patch bother about just the technical comments IMO.

amarts commented:
With a small change Gopkg.toml to point to "v4.1.0", instead of master, i could get further with errors, but below is the next hurdle.

```
$ make
....
go build -o build/glusterfs-csi-driver  cmd/glusterfs/main.go
# github.com/gluster/gluster-csi-driver/vendor/k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1
vendor/k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1/zz_generated.conversion.go:39:15: scheme.AddGeneratedConversionFuncs undefined (type *runtime.Scheme has no field or method AddGeneratedConversionFuncs)
# github.com/gluster/gluster-csi-driver/vendor/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1
vendor/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1/zz_generated.conversion.go:39:15: scheme.AddGeneratedConversionFuncs undefined (type *runtime.Scheme has no field or method AddGeneratedConversionFuncs)
# github.com/gluster/gluster-csi-driver/vendor/k8s.io/kubernetes/pkg/util/parsers
vendor/k8s.io/kubernetes/pkg/util/parsers/parsers.go:36:16: undefined: reference.ParseNormalizedNamed
make: *** [Makefile:8: glusterfs-csi-driver] Error 2
```

Madhu-1 commented on Gopkg.toml:
```diff
@@ -0,0 +1,28 @@
+
+[[constraint]]
+  name = "github.com/gluster/glusterd2"
+  version = "master"
```
need to change ```version = "master"``` to ```branch = "master" ```

harigowtham commented on pkg/glusterfs/controllerserver.go:
```diff
@@ -0,0 +1,353 @@
+	for _, vol := range volumes {
+		v, e := cs.client.VolumeStatus(vol.Name)
+		if e != nil {
+			return nil, status.Errorf(codes.Internal, "failed to get volume status %s", err.Error())
```
a listvolumes function erroring out for not being able to get the volume status while having the list of volumes is a bit off.

amarts commented:
Just an update on this patchset, as I commented earlier that @Madhu-1 and @humblec would work on this together, the updates to this patchset couldn't be posted to Humble's branch from Madhu. The code with review comments addressed, which can build fine, is committed to @Madhu-1's branch @ https://github.com/Madhu-1/gluster-csi-driver/commits/csi-driver (check 4 different commits)

@JohnStrunk @kshlm @prashanthpai @aravindavk @humblec @obnoxxx @nixpanic 

Request reviewers to have a look on those. The above patch didn't address the license issue, and as discussed in maintainer's meeting today, we can change the license later too, if author agree to it, and hence no issues there. @nigelbabu

humblec commented:
@JohnStrunk @kshlm @prashanthpai @aravindavk @humblec @obnoxxx @nixpanic @Madhu-1 Sorry for delay in addressing comments, it was not intentional rather due to natural calamities. :(.  The build locally passed due to "vendor" directory presence in my repo. Thanks @Madhu-1 for changing the dependency file for build issue in the other PR.  There are couple of things we can do now: 

*) I can push the docker image to gluster dockerhub and update the image location. 
*) I can pull the review comments commits from Madhu's PR and update this PR to have the history of review and other details and we can close https://github.com/gluster/gluster-csi-driver/pull/16

Let me know your thoughts @amarts @Madhu-1

amarts commented:
If you are fine with @Madhu-1 's changes, if we merge #16, this and that both will get closed with the same reference. So, no need to rebase again. Considering that patch is already reviewed by @kshlm and @aravindavk ... if you are ok with #16, feel free to merge it, so even this PR gets closed.

obnoxxx commented:
@amarts writes:
> If you are fine with @Madhu-1 's changes, if we merge #16,  this and that both will get closed with the same reference. So, no need  to rebase again. Considering that patch is already reviewed by @kshlm and @aravindavk ... if you are ok with #16, feel free to merge it, so even this PR gets closed.

Amar, how does that go with your previous statement (https://github.com/gluster/gluster-csi-driver/pull/11#issuecomment-413425327) that supported the completion the heketi-based PR #1 and build the gd2 based one on top as a configurable mode?

obnoxxx commented:
Previously, some reasons for not not merging the PR #1 for months were lack of infrastructure and CI etc. Has something changed? Why does everyone now seem to be in a hurry to merge one of the gd2-based drivers without a proper plan (as it seems) for the heketi based one and without any CI / tests in this repo?

amarts commented:
@JohnStrunk @obnoxxx 

> ANY CSI driver in this repo should be considered completely unsuitable for use without:
> 
> Proper documentation
> Automated container builds
> 

IMO, we need to get this PR merged, to get the above 2 done, mainly because it is worked from more than 1 person and teams. We surely need all these for calling it usable for sure.

> Automated, per PR unit and integration testing

IMO, For merging a PR, if we keep 'automation' as the goal, I am not sure we can release any tech-preview/experimental features, faster. My personal take is, we should make a release when we all feel confident. Meantime code in a repo should continue in a constant improvement curve considering the previous point, it can be multiple PRs. Happy to hear others' take here.

About PR #1 (gd1/heketi) based CSI driver:
@obnoxxx again, it is my personal opinion that,  GD1 /heketi based CSI, and GD2 based CSI need not be same. That way we consider only #11 or #16 for  [gcs](github.com/gluster/gcs), and PR #1 can continue to exist for people using current stack. This is my take as CSI is currently providing just 2 primitive APIs, and bringing that switch option would make it bulkier than it needs to be. Happy to hear other opinions.

obnoxxx commented:
> @JohnStrunk @obnoxxx
> >  **ANY** CSI driver in this repo should be considered completely unsuitable for use without:
>  >
> > * Proper documentation
> > * Automated container builds
> 
> IMO, we need to get this PR merged, to get the above 2 done, mainly because it is worked from more than 1 person and teams. We surely need all these for calling it usable for sure.

I don't get this argument, really. :-)

For me, at the very least, we need to establish some kind of travis ci jobs that calls some build and selftest make targets (or so) to have a minimal validation. It builds (even a simple go build would be a start, container build of course better), it passes some basic tests, etc...

I see no contradiction in introducing those tests with this PR or even before.

We should stop making the same horrible mistakes that we have been doing in other projects (starting without tests etc). This is a new repo. This is a chance to do it right from the start.

> > * Automated, per PR unit and integration testing
>
> IMO, For merging a PR, if we keep 'automation' as the goal, I am not sure we can release any tech-preview/experimental features, faster.

IMHO, "faster" is a lame excuse for being lazy: If you want to get stuff merged faster, but without the scrutiny of tests and verification, that means you want to add (possibly) broken code that you will have to fix later. So the "faster" aspect is actually deceiving. We did it with gd2 before, and that's been a bad choice, in my opinion.

>  My personal take is, we should make a release when we all feel confident. Meantime code in a repo should continue in a constant improvement curve considering the previous point, it can be multiple PRs. Happy to hear others' take here.

We are not even talking about a release. We are talking about a minimal barrier of quality and verification that we need. I don't want barrier-less merges. They lead to bad and broken code and more technical debt. Implementing such build+test gates is not a punishment, and not rocket-science. It does not inhibit progress or improvements, but actually fosters that.

> About PR #1 (gd1/heketi) based CSI driver:
> @obnoxxx again, it is my personal opinion that, GD1 /heketi based CSI, and GD2 based CSI need not be same.

I thought you were in agreement of having them the same (or at least to consider the possibility). Of course the don't *need* to be the same but, if you spare the work now, then you'll potentially have duplicate work later. Anyways, if we can have two drivers in this same repository (which I didn't think was possible), then I'm kind of ok. :-)

> That way we consider only #11 or #16 for gcs, and PR #1 can continue to exist for people using current stack. This is my take as CSI is currently providing just 2 primitive APIs, and bringing that switch option would make it bulkier than it needs to be. Happy to hear other opinions.

Yeah, it's a tradeoff. But as said, if we can have more than one driver here -- fine. Let's hear others.

humblec commented:
Let me share my thoughts here, first of all, indeed there was an urgency to get this moving or initial prototype ready without much delay to meet GCS product timeline. That was one of the main reason for me to push the PR in hurry and that too at non business hours/weekend :) . I was keep polishing the source code with various enhancements but waiting for everything to land in initial PR was not asked for.. As an alternative,  I have opened  issues in this repo which I think need to be addressed without much delay and tagged/labelled as `mvp 1.0` and E2E tests are part of it.  Deadline Vs completeness has a tradeoff and my thought was that, atleast as a team we are in agreement.  

Reg# PR#1 or heketi GD1 CSI driver, the immediate focus was shifted to GD2 based CSI driver after we reached to consensus on which API ( GD2 API) to use with csi driver in our planning meetings, Is there any change in this decision which I am not aware of ?  it does not mean, the heketi + gd1 driver shouldnt be there, we could have that as well. But IMO, we dont want to club both into same sources, rather we can keep it different.

What other enhancements can be done here?

If the ask is for complete E2E tests on this with this first PR, we need time especially when we are working on this on non business hours or weekend. If the ask is about an automatic container build with each version change we can come up with a deploy script and may be a travis build. It does not mean, the e2e tests are not part of initial `product`, the issues are prioritized with github labels and we can make sure those are addressed when we announce tech preview version of this driver.

obnoxxx commented:
@humblec writes:
> Let me share my thoughts here, first of all, indeed there was an urgency to get this moving or initial prototype ready without much delay to meet GCS product timeline. That was one of the main reason for me to push the PR in hurry and that too at non business hours/weekend :) .

Humble, don't get me wrong: I don't criticize the presence of the PR. It is good to have it! I question the urge to merge it too quickly. Quite honestly, we should not sacrifice all good manners and approaches in the upstream projects to meet some random downstream/product driven deadlines. Let's do this right. If it takes more time then so be it. I'm willing to fight this through. I've always said those timelines are too ambitious... But let's ignore the downstream topic here. We should not make compromises in the quality of this new upstream project. We have been doing it too often already. And it will come back with revenge if we compromise now. I just want us to do it right and well in the upstream. And I think @JohnStrunk is with me on this.

> I was keep polishing the source code with various enhancements but waiting for everything to land in initial PR was not asked for.. As an alternative, I have opened issues in this repo which I think need to be addressed without much delay and tagged/labelled as mvp 1.0 and E2E tests are part of it. Deadline Vs completeness has a tradeoff and my thought was that, atleast as a team we are in agreement.

I am not. And what deadlines are we speaking about, anyways? Is it a downstream/product deadline? Then it's irrelevant here at this point. If it is an upstream deadline, where is it written down? And anyways, it is a mistake to have an upstream deadline for a project that is just about to be created... I stand my statement: no compromises on quality beyond a certain minimum standard.

> Reg# PR#1 or heketi GD1 CSI driver, the immediate focus was shifted to GD2 based CSI driver after we reached to consensus on which API ( GD2 API) to use with csi driver in our planning meetings, Is there any change in this decision which I am not aware of ? it does not mean, the heketi + gd1 driver shouldnt be there, we could have that as well. But IMO, we dont want to club both into same sources, rather we can keep it different.

As said before, that's fine. If everyone is in favor of 2 separate drivers, and that's easily possible in this repo, then no issues! I have always been in favor of finishing the gd1/heketi based csi driver. Because this is something that is already used in the communities and in production in hundreds of setups. So this is a setup well worth covering!

> What other enhancements can be done here?
> 
> If the ask is for complete E2E tests on this with this first PR, we need time

Well complete e2e tests, I don't know. We need to sort it out. See @JohnStrunk 's comment.

Minimum:

* it builds (go build, and ideally container build)
* some additional tests (go fmt, go vet, shellcheck, ....)
* at least some unit/subsystem tests (at the api level, etc, maybe with some mocking)
* ideally some functional/e2e/blackbox tests either directly through the API or through kubernetes.

The last point is kind of optional in my opinion for a first cut, but can't be delayed long.
And yes. This will need some time. We need to take that time, imho. We in the upstream project can define when we are ready. I think we are not ready yet.

>  especially when we are working on this on non business hours or weekend.

Let's not bring this topic into the upstream discussion too much, but let me say that while personal time efforts are of course helpful sometimes, and the personal engagement is highly appreciated, it is certainly not a requirement!!

> If the ask is about an automatic container build with each version change we can come up with a deploy script and may be a travis build. It does not mean, the e2e tests are not part of initial product,

Let's not speak product. Product is downstream. Let's speak upstream project and releases at this stage. I think the main point is that the minimal bar for letting code into this new project should include the above.

Let's discuss more, maybe in person, and in a group if we can not resolve this here. :-)

> the issues are prioritized with github labels and we can make sure those are addressed when we announce tech preview version of this driver.

aravindavk commented:
Basic tests infra PR sent to #16, which tests build, fmt and a few other tests. `golint` tests are disabled for now because of conflict between CSI spec and `golint`.(For example: `./pkg/glusterfs/nodeserver.go:121:23: method NodeGetId should be NodeGetID`)

All tests infra taken from `gluster/glusterd2`

Need help from @kshlm, @prashanthpai to trigger `extras/centos-ci.sh` for every patch.

obnoxxx commented:
@aravindavk writes:
> Basic tests infra PR sent to #16, which tests build, fmt and a few other tests. golint tests are disabled for now because of conflict between CSI spec and golint.(For example: ./pkg/glusterfs/nodeserver.go:121:23: method NodeGetId should be NodeGetID) All tests infra taken from gluster/glusterd2 Need help from @kshlm, @prashanthpai to trigger extras/centos-ci.sh for every patch.

Can we possibly use travis for those lightweight tests, like we do in `heketi`. centos-ci should imho be used for more heavyweight tests that require vms.

obnoxxx commented:
@aravindavk and I forgot to say: cool, thanks for adding tests! :-)

humblec commented:
We can include https://github.com/gluster/gluster-csi-driver/issues/17 as well in the PR.

obnoxxx commented:
Huh, this has been merged, even though there were multiple outstanding change requests, and now I don't even see a merge commit in the branch. This is confusing. I guess what happened was that pr #16 was actually rebased on top of this one, and so this merge was a no-op. Still confusing.

I wished we had disabled merge commits before starting to merge PRs and had enforced a clean history from the beginning. :-)
